# Stats 21: Homework 3 + 4 (counts for two due to length)

## Takao Oba

Acknowledgements: several of these problems are copied from or modified from Think Python by Allen Downey.

I've started the homework file for you. You'll need to fill in the rest with your answers.
My encouragement is to use the keyboard shortcuts as much as possible and use the mouse as little as possible while working the Jupyter Notebook.

After you complete the homework with your answers, go to the menu and choose **Kernel > Restart & Run All.** Review the document to **make sure all requested output is visible**. You will not get credit for problems where the requested output is not visible, even if the function you coded is correct.

When you are satisfied with the output, choose File > Download As ... > PDF or  HTML. If you choose to save as HTML, you'll then need to "Print as PDF". Submit the PDF to Gradescope.

Submit this ipynb file, complete with your answers and output to Canvas / Bruin Learn.

**Again, you must make sure all requested output is visible to receive full credit.**

# Problem 1

An important part of programming is learning to interpret error messages and understanding what correction needs to be made.

Read and familiarize yourself with the following error messages.

Explain the error. Then duplicate each cell and correct the error. The first problem has been done for you as an example.

In [1]:
# A
print "Hello World"

SyntaxError: Missing parentheses in call to 'print'. Did you mean print("Hello World")? (910743254.py, line 2)

Answer: The `print()` function is missing parentheses. This results in a syntax error.

In [2]:
# corrected:
print("Hello World")

Hello World


In [4]:
# B
print("Hello")
    print("Goodbye")

IndentationError: unexpected indent (66443029.py, line 3)

Answer: The second line is indented when it should not need to be. This result is a indentation error. Indenting is commonly used for lines inside the function definition or inside loops or condition statements.

In [6]:
# C
x = 10
if x > 8
    print("x is greater than 8")

SyntaxError: invalid syntax (4211757722.py, line 3)

Answer: The line `if x > 8` is missing a colon at the end of the line. The result is a syntax error.

In [7]:
# D
if x = 10:
    print("x is equal to 10")

SyntaxError: invalid syntax (2686327292.py, line 2)

Answer: Single equal sign is used for assignment and to check for an equal condition, double equal sign `==` must be used. The result is a syntax error. 

In [8]:
# E
x = 5
if x == 5:
print("x is five")

IndentationError: expected an indented block (2346592504.py, line 4)

Answer: The last line should be indented because it is part of the result when the if statement is met. It is part of the if condition block. The result is an indentation error.

In [9]:
# F
l = [1, 2, 50, 10]
l = sort(l)

NameError: name 'sort' is not defined

Answer: There is a difference between sort and sorted. In this case sorted() should be used because sort is attached after an object after a period so for example l.sort(). The result is a name error.

# Problem 2

Use Python as a calculator. Enter the appropriate calculation in a cell and be sure the output value is visible.

A. How many seconds are there in 42 minutes 42 seconds?

In [10]:
42 * 60 + 42

2562

There are 2562 seconds in 42 minues and 42 seconds

B. There are 1.61 kilometers in a mile. How many miles are there in 10 kilometers?

In [11]:
10 / 1.61

6.211180124223602

There are 6.211180124223602 miles in 10 kilometers

C. If you run a 10 kilometer race in 42 minutes 42 seconds, what is your average 1-mile pace (time to complete 1 mile in minutes and seconds)? What is your average speed in
miles per hour?

In [18]:
# The average 1-mile pace 
kilo_to_mile = 10 / 1.61
min_to_sec = 42 * 60 + 42

print(min_to_sec / kilo_to_mile / 60) # this is minutes, but seconds is over 100 not over 60
print(((min_to_sec / kilo_to_mile / 60) % 1) * 60 / 100) #this is seconds over 60

6.874700000000001
0.5248200000000004


The Average 1 miles pace is 6 minutes and 52 seconds

In [20]:
# The average speed in miles per hour
sec_to_min = 42 / 60
sec42 = 42 + sec_to_min
sec_to_hour = sec42 / 60
10 / 1.61 / sec_to_hour

8.727653570337614

The average speed in miles per hour is 8.72765 miles per hour.

# Problem 3
Write functions for the following problems.

A. The volume of a sphere with radius r is $$V = \frac{4}{3}\pi r^3$$

Write a function `sphere_volume(r)` that will accept a radius as an argument and return the volume.

- Use the function to find the volume of a sphere with radius 5.
- Use the function to find the volume of a sphere with radius 15.

In [26]:
import math
def sphere_volume(r):
    out = (4/3) * math.pi * r**3
    print(out)

In [27]:
sphere_volume(5)

523.5987755982989


In [28]:
sphere_volume(15)

14137.166941154068


B. Suppose the cover price of a book is \\$24.95, but bookstores get a 40\% discount. Shipping costs \\$3 for the first copy and 75 cents for each additional copy.

Write a function `wholesale_cost(books)` that accepts an argument for the number of books and will return the total cost of the books plus shipping.

- Use the function to find the total wholesale cost for 60 copies.
- Use the function to find the total wholesale cost for 10 copies.

In [31]:
def wholesale_cost(books):
    discount_price = 24.95 * 0.6
    if books == 1:
        print(discount_price*books + 3)
    elif books > 1:
        print(discount_price*books + 0.75*(books - 1) + 3)

In [32]:
wholesale_cost(60)

945.4499999999999


In [33]:
wholesale_cost(10)

159.45


C.  Can you program a simple square root function without importing a module/library/package?

In [2]:
def square_root(num):
    return num ** 0.5

In [5]:
square_root(100)

10.0

# Problem 4

Use `import math` to gain access to the math library.

Create a function `polar(real, imaginary)` that will return the polar coordinates of a complex number.

The input arguments are the real and imaginary components of a complex number. The function will return a tuple of values: the value of the radius `r` and the angle `theta`. 

For a refresher, see: https://ptolemy.berkeley.edu/eecs20/sidebars/complex/polar.html

Show the results for any one the following complex numbers:

- 1 + i
- -2 - 3i
- 4 + 2i

In [7]:
import math

def polar(real, imaginary):
    theta = math.atan(imaginary / real)
    r = math.sqrt(real**2 + imaginary**2)
    print("theta = " + str(theta) + ", r = " + str(r))

In [8]:
polar(1,1)

theta = 0.7853981633974483, r = 1.4142135623730951
theta = 0.46364760900080615, r = 4.47213595499958


# Problem 5

Define a function called `insert_into(listname, index, iterable)`. It will accept three arguments, a currently existing list, an index, and another list/tuple that will be inserted at the index position.

Python's built-in function, `list.insert()` can only insert one object.

In [13]:
# write your code here
def insert_into(listname, index, iterable): 
    for i in iterable:
        listname.insert(index, i)
        index += 1
    return(listname)

In [14]:
# do not modify. We will check this result for grading
l = [0,'a','b','c',4,5,6]
i = ['hello', 'there']
insert_into(l, 3, i)

[0, 'a', 'b', 'hello', 'there', 'c', 4, 5, 6]

# Problem 6

Define a function called `first_equals_last(listname)`

It will accept a list as an argument. It will return `True` if the first and last elements are equal and the if the list has a length greater than 1. It will return False for all other cases.

In [59]:
# write your function here
def first_equals_last(listname):
    if len(listname) < 1:
        print("False")
    else:
        if listname[0] == listname[-1]:
            print("True")
        else:
            print("False")

In [60]:
# do not modify. We will check this result for grading
a = [1,2,3]
first_equals_last(a)

False


In [61]:
# do not modify. We will check this result for grading
b = ['hello','goodbye','hello']
first_equals_last(b)

True


In [57]:
# do not modify. We will check this result for grading
c  = [1,2,3,'1']
first_equals_last(c)

False


In [16]:
# do not modify. We will check this result for grading
d = [[1,2],[3,2],[1,2]]
first_equals_last(d)

NameError: name 'first_equals_last' is not defined

In [25]:
a = 1,2
b = 1,4
b -a

TypeError: unsupported operand type(s) for -: 'tuple' and 'tuple'